Listen to udev "add" events to detect devices that have been added and bind mount these within pod's containers.

amshinde opened the pull request:
From within a container, if a device is attached to a user driver, resulting in new device nodes being created under /dev, these are not visible from within a container. This is because /dev is mounted using tmpfs in a separate mount namespace. We need to continue using tmpfs, to restrict access to the initial set of devices available to the container. But we do need to provide a way for the container to access devices created later on based on a whitelist/blacklist.

This PR fixes this by watching for devices created and bind mounts these to each container's /dev so that the container is able to access the devices.

amshinde committed:
dev: Mount /dev from tmpfs before mount namespace is created

We use tmpfs to mount a container's /dev so that the device gets
access to a limited set of devices, but this means that the container
cannot access devices that have been created within the container.

We need to mount newly created devices to a container's /dev so that
the container can get access to these devices.

/dev was being mounted by libcontainer in container's mount namespace
followed by a chroot into the container's rootfs.
As this is a newly created mount point, it is not possible to access
a container's /dev from the agent mount namespace.

Even if we do a setns using something like docker's nsenter cgo package,
this would mean we would not be able to access the outside devices
because of the chroot created before the libcontainer call for starting
a container returns.

To overcome this, set up /dev for a container within the agent so
that the mount point is shared with the agent's mount namespace.
Note, this still uses tmpfs so that we get an initial limited set of
devices within the container.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde committed:
dev: Listen to udev events and bind mount these to container's /dev

Have a goroutine listening to udev "add" events and bind mount
the device node added to the pod's containers.
Do not handle block devices as they are handled elsewhere and
we do not want to add these.

Fixes #179

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

jodh-intel commented:
CI is barfing on lint errors.

jodh-intel commented on agent.go:
```diff
@@ -480,6 +488,69 @@ func (p *pod) closeChannels() {
+				continue
+			}
+
+			// Add subsystems that need to be blacklisted here
```
I'd add in code to support blacklisting now. The list might be empty, but it will then be easier to support future blacklisting, but also it will allow this code to be tested.

Speaking of tests, can you add some? If unit tests aren't possible/realistic, can you create a set of tests in https://github.com/clearcontainers/tests somewhere to exercise this new functionality?

amshinde replied:
@jodh-intel See https://github.com/clearcontainers/tests/issues/794.

amshinde replied:
I have added in the code to support blacklisting.

jodh-intel commented on agent.go:
```diff
@@ -480,6 +502,70 @@ func (p *pod) closeChannels() {
+				continue
+			}
+
+			if isBlacklistedSubsystem(d.Subsystem()) {
```
Thanks! Just one other comment - it might be useful to add a log call when a device is being black-listed.

sameo commented on agent.go:
```diff
@@ -480,6 +502,70 @@ func (p *pod) closeChannels() {
+				"udev-event": d.Action(),
+			}).Info("Received udev event")
+
+			// Ignore udev events for block devices, these are handled by another go-routine
```
Which routine handle block devices?

mcastelino replied:
@amshinde why all the containers. Should it not be based on per container device assignment. At this point the device list and mounts come per container config.json right

/cc @sboeuf

amshinde replied:
When the device node is created with the user interacting on the container command line, it is not possible to know which container the device node should be created in. But since all containers in a pod are supposed to share the same namespaces for k8s, I think this behaviour should be ok.

sameo commented on agent.go:
```diff
@@ -480,6 +502,70 @@ func (p *pod) closeChannels() {
+
+			if d.Action() == "add" {
+				// Bind device to /dev within all containers
+				p.bindDeviceNode(d.Devnode())
```
Should we unbind those devices when deleting the container?

amshinde replied:
@sameo I have done all the unbinding of devices including /dev. 
But I am seeing failures with CRIO.

sameo commented on agent.go:
```diff
@@ -1161,6 +1161,11 @@ func newContainerCb(pod *pod, data []byte) error {
 	}
 
+	dest := filepath.Join(absoluteRootFs, "/dev")
+	if err := syscall.Mount("tmpfs", dest, "tmpfs", syscall.MS_NOSUID|syscall.MS_STRICTATIME, ""); err != nil {
```
Same here, I think we should umount that one when deleting the container.

mcastelino replied:
@amshinde @sboeuf should we not setup a /dev directory per container. Not all containers may be allowed to see all mount points. Maybe I am reading this wrong and this is what you are already doing.

sboeuf replied:
As discussed offline, this should be gated by the fact that the user has requested a mount `/dev:/dev`. This is the way we would do that with runc, and this is the proper way to make sure the user expects to see new devices being created and to be able to get access to them.

sboeuf replied:
This raises the question of handling special cases like this one, where cannot expect to mount the directory from the host. Should we maintain a list of such directories ? Actually, we should define such a list first and see how we plan to handle every case.

amshinde replied:
@sboeuf See this commit where I have addressed `-v /dev:/dev` : https://github.com/clearcontainers/agent/pull/180/commits/0680f5a626c85ad5daf299e2bc1903b945ce1a5d

mcastelino commented on agent.go:
```diff
@@ -480,6 +504,80 @@ func (p *pod) closeChannels() {
+// listenToUdevEvents listens to udev events to detect device nodes added
+// and binds the device nodes to the /dev of all the containers.
+// This is required as a container has its /dev mounted with tmpfs with an initial
+// list of devices.
```
It would be good to call out when this happens. I assume this is for the use case where devices are specified as part of the container config.json and you are hot plugging them into the VM, and now expect them to show up in the specific container?

mcastelino replied:
Or is this due to a device creation due to some action in the container which has the CAPs to perform device creation, and this device is created due to an activity in the container itself does not show up in the container.

amshinde replied:
@mcastelino It is the latter case. When you are performing some action in the container that results in a device node being created in the VM. This is useful for running DPDK inside the container where you would typically bind a device to vfio or uio resulting in a new device node being created.

amshinde committed:
devices: Bind mount devices only when /dev is expected to be bindmounted

As per runc behaviour new devices are expected to show up when
-v /dev:/dev is passed on the docker command line. In this case, docker
passes /dev as a bind mount rather than a tmpfs based mount.

This really means that the host /dev should be replicated inside the container.
Since that is not possible in case of VM based container's, we do our
best by mounting the newly created devices while interacting with the
container.

We pass a bool "BindMountDev" to newContainer payload to flag this case. New
devices are bind-mounted only when this flag is true.
This flag needs to be passed by the runtime.

We also get rid of the VFIO bind mount, which was an interim solution.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde commented:
@sameo @sboeuf I pushed a few changes addressing `-v /dev:/dev` and unmounting of devices. Please have a look.

sboeuf commented:
thx @amshinde , I'll check that

amshinde commented:
@sboeuf I have introduced similar change in virtcontainers to pass information about the /dev bind mount:
https://github.com/containers/virtcontainers/pull/523

That needs to go in first.

sboeuf commented on agent.go:
```diff
@@ -1131,6 +1131,21 @@ func addMounts(config *configs.Config, fsmaps []hyper.Fsmap) error {
 	return nil
 }
 
+func mountDevtoRootfs(absoluteRootFs string) error {
```
s/`mountDevtoRootfs`/`mountDevToRootfs`/

sboeuf commented on agent.go:
```diff
@@ -1378,10 +1375,11 @@ func newContainerCb(pod *pod, data []byte) error {
+		container:    libContContainer,
+		config:       config,
+		processes:    processes,
+		bindMountDev: payload.BindMountDev,
```
The change looks okay to me, but same comment as for virtcontainers, this should be part of a more generic structure, so that container would have only one new field for any exceptions we have to make regarding bind mounts.

sboeuf reviewed (commented):
@amshinde about the "listen udev events" commit, why don't you merge the block device case with this case, that is having a global go routine listening constantly to uevents. You could provide a way to add more rules in order to make sure we don't miss any uevent you could expect.

amshinde commented:
@sboeuf I want to keep the two cases separate, as block devices are mounted based on the name passed on the command line.
```
docker run -it debian --device /dev/sdb:/dev/sdf.
```
Here the /dev/sdb passed to the VM may become "/dev/vda", we then need to mount that as "/dev/sdf" inside the container and not "/dev/vda". Hence we handle that separately.

sboeuf commented:
@amshinde ah yes that makes sense. There are so many cases to handle, I am worried about this complexity to handle many specific cases...

amshinde commented:
Yeah I know! Its just that many of the runc configs dont translate very well to us.

sboeuf commented:
@amshinde Yes agreed... Then please try to document into the code as much as you can so that it is easier to anyone who wants to take a look at it.

amshinde commented:
@sboeuf Will do.

amshinde commented:
@sboeuf I have included changes that I made to virtcontainers here

sboeuf commented:
LGTM

[![Approved with PullApprove](https://img.shields.io/badge/approvers-approved-brightgreen.svg)](https://pullapprove.com/clearcontainers/agent/pull-request/180/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent) [![Approved with PullApprove](https://img.shields.io/badge/reviewers-approved-brightgreen.svg)](https://pullapprove.com/clearcontainers/agent/pull-request/180/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

amshinde commented:
@sboeuf I want to vendor changes into the proxy before this goes in.

sboeuf commented:
Ok

amshinde commented:
@sboeuf I have vendored changes in proxy. Can you merge this if you are ok with this?

mcastelino reviewed (commented):
@amshinde can you clarify what you mean by

``We also get rid of the VFIO bind mount, which was an interim solution.``

amshinde commented:
@mcastelino before this PR, I had implemented a temporary solution of bind-mounting /dev/vfio directory within the container, so that devices bound to vfio can be seen.
With this PR we no longer require the temporary solution.
